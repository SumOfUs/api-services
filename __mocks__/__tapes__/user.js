module.exports = {
  email__startswith: {
    status: 400,
    data: {
      error: "'startswith' is not an allowed filter on the 'email' field.",
    },
  },
  'nonexistentuser@example.com': {
    data: {
      meta: {
        limit: 20,
        next: null,
        offset: 0,
        previous: null,
        total_count: 0,
      },
      objects: [],
    },
  },
  'example@example.com': {
    data: {
      meta: {
        limit: 20,
        next: null,
        offset: 0,
        previous: null,
        total_count: 1,
      },
      objects: [
        {
          actions: '/rest/v1/action/?user=388175',
          address1: '',
          address2: '',
          city: 'Philadelphia',
          country: 'United States',
          created_at: '2012-03-25T06:44:11',
          email: 'example@example.com',
          events: '/rest/v1/event/?creator=388175',
          eventsignups: '/rest/v1/eventsignup/?user=388175',
          fields: {
            en: '1',
          },
          first_name: 'Ashley',
          id: 388175,
          lang: '/rest/v1/language/100/',
          last_name: 'Mease',
          location: '/rest/v1/location/388175/',
          logintoken: '/rest/v1/user/388175/logintoken/',
          middle_name: '',
          orderrecurrings: '/rest/v1/orderrecurring/?user=388175',
          orders: '/rest/v1/order/?user=388175',
          phones: [],
          plus4: '',
          postal: '19145',
          prefix: '',
          rand_id: 692604506,
          region: '',
          resource_uri: '/rest/v1/user/388175/',
          source: 'starbucks 3/12',
          state: 'PA',
          subscription_status: 'bounced',
          subscriptionhistory: '/rest/v1/subscriptionhistory/?user=388175',
          subscriptions: '/rest/v1/subscription/?user=388175',
          suffix: '',
          token: '.388175.onPRSf',
          updated_at: '2017-01-31T00:07:52',
          usergeofields: '/rest/v1/usergeofield/?user=388175',
          usermailings: '/rest/v1/usermailing/?user=388175',
          useroriginal: '/rest/v1/useroriginal/388175/',
          zip: '19145',
        },
      ],
    },
  },
};
