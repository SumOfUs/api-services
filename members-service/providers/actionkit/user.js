// @flow weak
import { get } from 'axios-es6';
import { pick } from 'lodash';

// Accessing process.env is expensive, therefore we make
// a local copy of what we need
const environment = pick(
  process.env,
  'AK_API_URL',
  'AK_PASSWORD',
  'AK_USERNAME'
);

export function searchUser(filters, env = environment) {
  if (!env.AK_API_URL) return Promise.reject('AK_API_URL is not set');
  if (!env.AK_USERNAME || !env.AK_PASSWORD) {
    return Promise.reject('ActionKit credentials are not set');
  }
  return get(`${env.AK_API_URL}/user`, {
    auth: {
      username: env.AK_USERNAME,
      password: env.AK_PASSWORD,
    },
    params: { ...filters, limit: 1 },
  })
    .then(result => {
      return result.data.objects;
    })
    .catch(error => {
      if (error.status) {
        return Promise.reject({
          statusCode: error.status,
          body: error.data || '',
        });
      }
      return Promise.reject(error);
    });
}
