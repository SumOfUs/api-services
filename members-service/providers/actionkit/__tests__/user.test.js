// @flow
import { searchUser } from '../user';
import { pick } from 'lodash';

const env = pick(process.env, 'AK_API_URL', 'AK_PASSWORD', 'AK_USERNAME');

describe('searchUser', () => {
  test('returns an array of matches', () => {
    const params = { email: 'example@example.com' };
    return searchUser(params, env).then(result =>
      expect(result).toMatchSnapshot()
    );
  });

  test('returns an empty array if there are no matches', () => {
    const params = { email: 'nonexistentuser@example.com' };
    return searchUser(params, env).then(result =>
      expect(result).toMatchSnapshot()
    );
  });

  test('passes through the error responses, rejecting the promise', () => {
    return expect(
      searchUser({ email__startswith: true })
    ).rejects.toMatchSnapshot();
  });

  test('rejects with an error message if AK_API_URL is not set', () => {
    return expect(
      searchUser(null, { ...env, AK_API_URL: '' })
    ).rejects.toMatchSnapshot();
  });

  test('rejects with an error message if AK_USERNAME is not set', () => {
    return expect(
      searchUser(null, { ...env, AK_USERNAME: '' })
    ).rejects.toMatchSnapshot();
  });

  test('rejects with an error message if AK_PASSWORD is not set', () => {
    return expect(
      searchUser(null, { ...env, AK_PASSWORD: '' })
    ).rejects.toMatchSnapshot();
  });
});

// Mock AK API requests.
// To make real requests (and update snapshots), comment the following lines:
jest.mock('axios-es6', () => ({
  get(query, options) {
    if (options.params.email) {
      return Promise.resolve(
        require('../__tapes__/searchUser')[options.params.email]
      );
    }

    if (options.params.email__startswith) {
      const error = require('../__tapes__/searchUser')['email__startswith'];
      return Promise.reject(error);
    }

    return Promise.reject(new Error('connect ECONNREFUSED 127.0.0.1:80'));
  },
}));
