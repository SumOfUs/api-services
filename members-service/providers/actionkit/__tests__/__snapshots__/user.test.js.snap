// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searchUser passes through the error responses, rejecting the promise 1`] = `
Object {
  "body": Object {
    "error": "'startswith' is not an allowed filter on the 'email' field.",
  },
  "statusCode": 400,
}
`;

exports[`searchUser rejects with an error message if AK_API_URL is not set 1`] = `"AK_API_URL is not set"`;

exports[`searchUser rejects with an error message if AK_PASSWORD is not set 1`] = `"ActionKit credentials are not set"`;

exports[`searchUser rejects with an error message if AK_USERNAME is not set 1`] = `"ActionKit credentials are not set"`;

exports[`searchUser returns an array of matches 1`] = `
Array [
  Object {
    "actions": "/rest/v1/action/?user=388175",
    "address1": "",
    "address2": "",
    "city": "Philadelphia",
    "country": "United States",
    "created_at": "2012-03-25T06:44:11",
    "email": "example@example.com",
    "events": "/rest/v1/event/?creator=388175",
    "eventsignups": "/rest/v1/eventsignup/?user=388175",
    "fields": Object {
      "en": "1",
    },
    "first_name": "Ashley",
    "id": 388175,
    "lang": "/rest/v1/language/100/",
    "last_name": "Mease",
    "location": "/rest/v1/location/388175/",
    "logintoken": "/rest/v1/user/388175/logintoken/",
    "middle_name": "",
    "orderrecurrings": "/rest/v1/orderrecurring/?user=388175",
    "orders": "/rest/v1/order/?user=388175",
    "phones": Array [],
    "plus4": "",
    "postal": "19145",
    "prefix": "",
    "rand_id": 692604506,
    "region": "",
    "resource_uri": "/rest/v1/user/388175/",
    "source": "starbucks 3/12",
    "state": "PA",
    "subscription_status": "bounced",
    "subscriptionhistory": "/rest/v1/subscriptionhistory/?user=388175",
    "subscriptions": "/rest/v1/subscription/?user=388175",
    "suffix": "",
    "token": ".388175.onPRSf",
    "updated_at": "2017-01-31T00:07:52",
    "usergeofields": "/rest/v1/usergeofield/?user=388175",
    "usermailings": "/rest/v1/usermailing/?user=388175",
    "useroriginal": "/rest/v1/useroriginal/388175/",
    "zip": "19145",
  },
]
`;

exports[`searchUser returns an empty array if there are no matches 1`] = `Array []`;
